### PR TITLE
feat(ai): server-side auto-retry for agent loops that never finish

### PR DESCRIPTION
**Pull request by 2witstudios**
## Problem

Agents sometimes run a long string of tool calls and then **never produce a final message** — most often with **OpenRouter**, where the provider connection drops mid-loop. Today this is invisible:

- `streamText` has no `onError` and `createUIMessageStream` has no `onError`, so a mid-stream provider disconnect is **masked** as a generic `"An error occurred."` part.
- `onFinish` then saves the partial assistant message and marks the stream `complete`, so a broken run looks clean to everything downstream.
- The client `useStreamRecovery` only auto-retries *browser-connection* drops, and the masked generic error doesn't match — so nothing retries.

## Approach — server-side, in-request retry (functional core / imperative shell)

- **`agent-finish-classifier.ts`** — pure `classifyAttempt()`, total over `FinishReason` → `clean` / `retry` / `terminal`. No IO; exhaustively unit-tested.
- **`run-agent-with-retry.ts`** — bounded retry loop **inside `createUIMessageStream.execute`**, so `onFinish` still fires **exactly once** → one credit hold / one settle (no double-charge), with steps **summed across attempts** so we bill the real provider cost. One message envelope. Conservative budget: **2 retries, 0.5/1.5s backoff, wall-clock cap under the 300s function limit**.
- **`provider-factory.ts`** — OpenRouter now sends `extraBody.provider.require_parameters: true` (+ `allow_fallbacks`), so tool-calling never routes to a provider that silently ignores `tools` — a **root cause** of "never finishes". Keeps `usage:{include:true}`.
- Both streaming routes (`api/ai/chat`, `api/ai/global/[id]/messages`) wired through a shared `buildStreamText` factory.

## The hard constraint: append-only streams (the core correctness guard)

The UI message stream is **append-only** — a second `start` chunk does *not* reset accumulated parts (verified in the SDK). So a from-scratch retry of an attempt that **already streamed content** would concatenate duplicate/garbled output into the persisted+multicast message, orphan a `text-start`, and re-run a mutating tool whose side effect already ran.

The guard, applied at the source — **never re-stream content that's already on the wire:**

- The pipe reports the first content chunk (`onContent`, fires even if the stream then throws).
- An attempt is **only retried when it streamed *zero* content** (a pure pre-first-token failure: provider 5xx/429/no-tool-capable-provider/immediate disconnect). Since tool-call/result chunks *are* content, this also means **no retry ever re-runs a tool** — the double-execution risk is structurally impossible, so each retry safely re-runs the original messages with no re-feed.
- A failure **after** any content is **terminal** — surface a non-network error (the user/client retries); never duplicate.
- A `tool-calls` finish without the finish tool is terminal (step-budget, or an anomaly): under `stopWhen [hasToolCall(finish), stepCountIs(N)]` the SDK auto-continues every tool step, so a retry can't help.

So: pre-content failures recover transparently; mid-stream-after-content drops surface a clear error instead of duplicating; the `require_parameters` config independently prevents the no-tools-routing root cause.

## Observability + correctness in `onFinish`

- `success = (finalOutcome !== 'exhausted')` — a gave-up generation is no longer logged as success (cost still settles; the provider charged us).
- `metadata.retryAttempts / retryOutcome / retryTerminalReason`.
- `lifecycle.finish(aborted)` reflects a user stop during/after the loop (`terminalReason === 'aborted' || abortSignal.aborted`); idempotent with `onAbort`.

## Design decision

**Server-side only.** Deliberately did NOT add a client-side incomplete-detection path — it would compound retries and create two drifting definitions of "done". The existing `useStreamRecovery` network-retry (disjoint failure class) is untouched, and the server give-up error is phrased to *not* match its `isNetworkError` patterns.

## Correctness summary

- No double-billing / double-finish (loop inside `execute`; `lifecycle.finish` idempotent).
- **No content duplication / orphaned parts / double tool execution** — guaranteed by the never-re-stream-emitted-content guard (retry ⟹ zero content emitted ⟹ no tool ran).
- Misfire bounded to ambiguous-with-no-content (retried once).

## Tests

Classifier matrix incl. pre/post-content branches and the unreachable tool-calls cases; shell retry / cap / single-envelope / throw-after-content→terminal; OpenRouter config. **809 AI tests pass; typecheck + lint clean; CI green.**

## Pre-merge checklist

- [ ] Manual e2e: force an OpenRouter disconnect *before* first token (expect transparent recovery) and *after* content (expect a clean terminal error, no duplication).
- [ ] Verify the live `extraBody.provider` nesting against a real OpenRouter response.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * In-request retry loop for AI conversations with step caps, single start/finish stream envelope, and configurable backoff
  * Provider routing enhancements to improve model/tool selection

* **Bug Fixes**
  * Usage tracking now reflects real retry outcomes and reports success/failure accurately
  * Stream lifecycle now honors user aborts and better handles termination on stop

* **Tests**
  * Expanded test coverage for retry, finish classification, provider behavior, and streaming edge cases

### CODE REVIEW

File: apps/web/src/lib/ai/core/run-agent-with-retry.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    attempts = attempt + 1;
+    const aiResult = buildStreamText(messages);
+
+    let caughtError: unknown;
+    try {
+      await pipeUIMessageStreamStrippingStart(aiResult, writer, {
+        suppressStart: true,
+        suppressFinish: true,
+        suppressError: true,
+      });
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/core/run-agent-with-retry.ts:160]: Addressed in 9ff3e12b2 / fd53f5e9c via a guard rather than buffering. The pipe now reports the first content chunk (`onContent`, fires even if the stream then throws), and `classifyAttempt` only returns `retry` for a hard error/ambiguous finish when `emittedContent === false`. So once any text/tool chunk has been forwarded we never retry from scratch — the attempt becomes terminal and we surface a non-network error instead of duplicating. Because tool-call/result chunks are themselves content, a retried attempt provably ran no tools, so each retry safely re-runs the original messages with no re-feed. Buffering would have cost live streaming for the common success-first case; this preserves it while making mid-content duplication impossible.

File: apps/web/src/lib/ai/core/run-agent-with-retry.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+  }
+
+  // Surface a TERMINAL, non-network error on provider-error give-up. Phrasing avoids the
+  // client's isNetworkError patterns so useStreamRecovery does not re-run on top of us.
+  if (finalOutcome === 'exhausted' && terminalReason === 'provider-error') {
+    safeWrite({
+      type: 'error',
+      errorText:
+        'The assistant could not complete its response after several attempts. Please try again.',
+    });
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/core/run-agent-with-retry.ts]: Fixed in fd53f5e9c. The give-up error is now surfaced on any `finalOutcome === 'exhausted'` (which covers provider-error AND ambiguous/time-budget exhaustion — all of which produced no content) as well as the after-content provider-error terminal. Content-bearing terminals (length / content-filter / step-budget / tool-calls-no-finish) and user aborts still close normally since they already streamed a real response. Note: the `tool-calls-no-finish` *retry* path referenced here was removed entirely — under `stopWhen [hasToolCall(finish), stepCountIs(N)]` the SDK auto-continues every tool step, so that outcome is unreachable below the cap and is now classified terminal. Added a test for ambiguous-exhaustion → error.

File: apps/web/src/lib/ai/core/run-agent-with-retry.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    const finishReason: FinishReason | undefined = await aiResult.finishReason.catch(
+      () => undefined,
+    );
+    const responseMessages: ModelMessage[] = await aiResult.response
+      .then((r) => r.messages)
+      .catch(() => []);
+    const steps: ProviderMetadataCarrier[] = await aiResult.steps
+      .then((s) => s as ProviderMetadataCarrier[])
+      .catch(() => []);
+    const usage = await aiResult.totalUsage.catch(() => undefined);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/core/run-agent-with-retry.ts]: Fixed in 74649290d. `buildStreamText(baseMessages)` is now invoked inside the attempt try/catch, so a synchronous factory throw is captured as `caughtError` and classified as provider-error (retryable when no content) rather than escaping `runAgentWithRetry`. The per-attempt field reads (`finishReason`/`response`/`steps`/`totalUsage`) guard on a possibly-undefined `aiResult`, and the single start/finish envelope is preserved. Added a test covering sync-throw → retry → clean.

File: apps/web/src/lib/ai/core/run-agent-with-retry.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+      reason: outcome.reason,
+      elapsedMs: elapsed,
+    });
+
+    await new Promise((resolve) => setTimeout(resolve, backoffMs(attempt)));
+    if (abortSignal.aborted) {
+      finalOutcome = 'terminal';
+      terminalReason = 'aborted';
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/core/run-agent-with-retry.ts]: Fixed in 74649290d. The backoff now races the timer against an `abort` listener (clearing the timer on abort), so it resolves immediately when the user stops mid-wait instead of holding the request open for the full delay. The post-backoff `abortSignal.aborted` check then breaks with terminalReason 'aborted'.
